### DISCUSSION TITLE
Level meter for audio clips

### DISCUSSION THREAD

**Discussion by Psirus** [SynthstromAudible/DelugeFirmware#259] [Ideas]
When in clip view on an audio clip, the mute & audition columns could act as a "pre-fader" level meter to enable better gain setting on the input. More than once I had a distorted sound simple because I was unknowingly driving the input too hard.

**Comment by m-m-adams** [comment]
This is a cool idea, do you want to try implementing it?

**Comment by Psirus** [comment]
In general, yes. But it won't be happening in the next days, maybe even weeks.

**Comment by m-m-adams** [comment]
No problem! Happy to help you through where/how to do the actual level calculations if you need it

**Comment by Psirus** [comment]
Any deluge/codebase specific hints are greatly appreciated. The level calculations in general I would like to research myself.

**Comment by m-m-adams** [comment]
You'll need to calculate the level in the audio clip render function in sound.cpp 
Then in the audio clip view you would periodically grab the level of the associated clip and display it on the pads. Note that the refresh rate will need to be low and that calculating the level will need to be done very efficiently to not cause voice dropouts - use multiple_accumulate_rshfift32 for the power calculation and then you can probably get away with using crappy_square_root since it'll just be displayed on 8-16 pads anyway

**Comment by wargoth** [comment]
I would like to vote for this feature. Input level monitoring is something Deluge lacks. I was thinking it would be nice to display input level with one of the gold knob LEDs, but displaying it on the mute/audition seems even better. Currently there is no way to say if the input signal is too low (noise floor too high) or clipping. 

Thank you!
